external-import/dragos/dragos/domain/use_cases/ingest_report.py

```diff
@@ -0,0 +1,241 @@
+        """Make an OCTI Observable and Indicator generator from Dragos report related indicators."""
+
+        def make_observable(related_indicator: "Indicator") -> "octi.Observable":
+            """Make an OCTI observable from a Dragos report related indicator."""
+            dragos_indicator_type = related_indicator.type.lower()
+            match dragos_indicator_type:
+                case "artifact":
+                    return self._make_artifact(related_indicator)
+                case "domain":
+                    return self._make_domain_name(related_indicator)
+                case "ip":
+                    return self._make_ip_address(related_indicator)
+                case "md5" | "sha1" | "sha256":
+                    return self._make_file(related_indicator)
+                case "url":
+                    return self._make_url(related_indicator)
+                case _:
+                    logger.warning(
+                        f"Unsupported indicator type {dragos_indicator_type}"
+                    )
+            return None
+
+        for related_indicator in report.related_indicators:
+            indicator = None
+            observable = make_observable(related_indicator)
+            if observable:
+                indicator = observable.to_indicator(
+                    valid_from=related_indicator.first_seen,
+                    valid_until=related_indicator.last_seen,
+                )
+            if observable and indicator:
+                yield (observable, indicator)
+
+    def make_indicator_based_on_observable_relationship(
+        self, indicator: "octi.Indicator", observable: "octi.Observable"
+    ) -> octi.IndicatorBasedOnObservable:
+        """Make an OCTI IndicatorBasedOnObservable relationship from Indicator and Observable."""
+        return octi.IndicatorBasedOnObservable(
+            author=self.author,
+            source=indicator,
+            target=observable,
+            markings=[self.tlp_marking],
+            # unused
+            description=None,
+            start_time=None,
+            stop_time=None,
+            external_references=None,
+        )
+
+    def make_report(
+        self, report: "Report", related_objetcs: list["octi.BaseEntity"]
+    ) -> octi.Report:
+        """Make an OCTI Report from a Dragos report and the related entities."""
+        return octi.Report(
+            name=report.title,
+            publication_date=report.created_at,
+            description=report.summary,
+            # labels=report.related_tags,
+            objects=related_objetcs,
+            author=self.author,
+            markings=[self.tlp_marking],
+            # unused
+            reliability=None,
+            report_types=None,
+            external_references=None,
+        )
+
+    def run_on(self, report: "Report") -> list["octi.BaseEntity"]:
+        """Run the process of entities creation thanks to a Report."""
+        entities: list["octi.BaseEntity"] = []
```

Comment:
due to yielded/returned None values

```suggestion
    def run_on(self, report: "Report") -> list[Optional["octi.BaseEntity"]]:
        """Run the process of entities creation thanks to a Report."""
        entities: list[Optional["octi.BaseEntity"]] = []
```

or if you use a filter before returning

 ```
    def run_on(self, report: "Report") -> list["octi.BaseEntity"]:
        """Run the process of entities creation thanks to a Report."""
        entities: list[Optional["octi.BaseEntity"]] = []

        ...

        entities.extend(self.make_domain_objects(report))

        filtered_entities = [entity for entity in entities if entity is not None]

        # Only append Report, Author and TLP if at least one entity is present
        # to prevent sending unconsistent bundle in application layer
        if filtered_entities:
            filtered_entities.append(self.make_report(report, filtered_entities))
            filtered_entities.append(self.author)
            filtered_entities.append(self.tlp_marking)

        return filtered_entities
 ```

Reply:
Now that `make_domain_objects` never returns `None`, `entities` list doesn't need to be filtered